### ISSUE TITLE
Slow response times from /api/v4/analytics/old happening on app open

### ISSUE THREAD

**Issue by DonCziken** [mattermost/mattermost-server#21674]
#### Summary
Recently we've been experiencing performance issues with our MM server, while the root cause was different, we've noted that the following request takes 600-700ms on initial load of the web and desktop app. This causes annoying entrance lag for admin users, while also produces a 403 error for non admin users on app open, both which we see as an issue.

`GET https://SERVER/api/v4/analytics/old?name=standard&team_id=`

This is how it look for admin user
![Screenshot from 2022-11-15 19-26-04](https://user-images.githubusercontent.com/36049985/202001200-aade697b-deda-477d-ae3d-828c73bc227c.png)

and that's how it looks for non admin user
![Screenshot from 2022-11-15 20-10-11](https://user-images.githubusercontent.com/36049985/202005111-e6910acc-2cc5-444b-b815-33119386eec9.png)

The endpoint returns statistics about MM usage, which are presented on the dashboard of System Console. The response size is tiny (few hundred bytes). It looks like there is some issue with calculating those. 

#### Steps to reproduce
1. Open MM web application with any browser
2. Authenticate and open any team/chat
3. Review debug tab to see network requests

#### Expected behavior
 a) request to be much faster, <100ms 
 b) request to not happen unless the system console is opened - as long as admin is not looking there this data is not presented anyway, also standard users are not expected to request for that data anyway.

#### Additional Info
Here are example contents of the response:
![Screenshot from 2022-11-15 19-28-26](https://user-images.githubusercontent.com/36049985/202002611-484967b1-4855-45d8-b32d-b00c724d1819.png)

Here is JS stack trace showing, that the call originate from `admin.ts` and done through `getAnalytics` call:
![Screenshot from 2022-11-15 19-29-11](https://user-images.githubusercontent.com/36049985/202006386-8444dd74-51d3-4b6f-889e-914ff71b9877.png)

**Comment by amyblais** [comment]
@DonCziken What Mattermost server version are you on?

**Comment by DonCziken** [comment]
@amyblais 

At the moment we are using:

Mattermost Version: 7.4.0
Database Schema Version: 93
Database: postgres (PostgreSQL) 14.4

licence: Mattermost Team Edition

**Comment by amyblais** [comment]
@agarciamontoro Does this look like something we could open a Jira ticket for to investigate further?

**Comment by agarciamontoro** [comment]
Thanks for the ping, @amyblais! And thank you very much for the detailed report, @DonCziken, we really appreciate it.

From your report, it seems there are two different issues here:
1. Why do we call `/analytics/old` on initial load, if that's only rendered in the system console? And why do we do it for regular users as well? I don't have the answer to that, but I created [this ticket](https://github.com/mattermost/mattermost-server/issues/21674) and assigned it to the Webapp Platform team to take a look. @hmhealey will know who the owner of this should be.
2. Why does a successful call to `/analytics/old` take 600-700ms? This is interesting. The way [it is implemented](https://github.com/mattermost/mattermost-server/blob/ab5137395cea997f78498caec9bbbe8af224125a/app/analytics.go#L33-L177) today, this time is not that weird for large databases, since we're doing a `SELECT COUNT(*) FROM` in several tables concurrently, so the request has to wait for the longest query, which will probably correspond to the largest table there is, which probably is `Posts`. In fact, that query is only executed if the number of users is under a threshold, `AnalyticsSettings.MaxUsersForStatistics`. So I have a couple of questions here before going further:
- What's the value of `AnalyticsSettings.MaxUsersForStatistics` in your config?
- Could you run `SELECT COUNT() FROM Posts` manually in your database and let me know both the result and the query's execution time?

**Comment by DonCziken** [comment]
@agarciamontoro thanks for looking at this and providing some initial feedback.

#### Regarding your summary on issues
yes indeed those are issues in this case.

#### Answer to your questions:
 a) `AnalyticsSettings.MaxUsersForStatistics` is 2.5k , however we have around 20 active users at the moment, so not sure if relevant?
 
 ```json
 "AnalyticsSettings": {
      "MaxUsersForStatistics": 2500
  },
  ```

 b) query to count posts take around 150ms, profiler shows as follows:

```sql
EXPLAIN ANALYZE select count(*) from Posts;
                                                                                QUERY PLAN                                                                                
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=21339.16..21339.17 rows=1 width=8) (actual time=143.734..143.916 rows=1 loops=1)
   ->  Gather  (cost=21338.94..21339.15 rows=2 width=8) (actual time=140.836..143.902 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial Aggregate  (cost=20338.94..20338.95 rows=1 width=8) (actual time=130.409..130.411 rows=1 loops=3)
               ->  Parallel Index Only Scan using idx_posts_is_pinned on posts  (cost=0.42..19476.66 rows=344915 width=0) (actual time=0.037..97.288 rows=266013 loops=3)
                     Heap Fetches: 181812
 Planning Time: 0.260 ms
 Execution Time: 144.053 ms
(9 rows)
```

#### Additional Response

>but I created *this link* and assigned it to the Webapp Platform

Can you fix the link here? as it points to the same ticket as this one.

I wonder if the stats should be calculated live each request, shouldn't there be some cache or schedule to calculate them? I mean that applies especially to db queries. 

Also personally I don't like the logic behind `AnalyticsSettings.MaxUsersForStatistics` at least in this case, as basically if I have more users than this threshold then I am not getting the posts statistic.

**Comment by agarciamontoro** [comment]
> we have around 20 active users at the moment

The number of users considered here is the output of `SELECT COUNT(DISTINCT u.Id) FROM Users AS u WHERE u.DeleteAt = 0`. How many user does that query return for you?

> query to count posts take around 150ms, profiler shows as follows:

Thanks for the `EXPLAIN ANALYZE`. Could you also let me know the result of the query? I'm interested in knowing how many posts your DB has. 144ms for scanning the whole table does not seem that much. Although your system is planning an index-only scan on `idx_posts_is_pinned`. That's weird :thinking: 

The only way to reduce that time would be to either `VACUUM` the table, which will take time (or `VACUUM FULL` to completely recreate it, which will take way more time and will block the table for the whole process), or increase your DB host specs. I just ran `VACUM FULL Posts` in my test DB with ~12 million posts and the query time got reduce from 700ms to 500ms, which is not that bad. If you can do it overnight, it may help a bit.

In any case, if this query takes 144ms, then it doesn't look it's the bottleneck of the `/analytics/old` endpoint in your system. Would it be possible for you to time all queries done for that endpoint (you can see them [here](https://github.com/mattermost/mattermost-server/blob/ab5137395cea997f78498caec9bbbe8af224125a/app/analytics.go#L33-L177)) and find out the one that takes the longest?

> Can you fix the link here? as it points to the same ticket as this one.

Sorry :facepalm: I edited the message, but here it is: https://mattermost.atlassian.net/browse/MM-48456

> I wonder if the stats should be calculated live each request, shouldn't there be some cache or schedule to calculate them? I mean that applies especially to db queries.

I agree that the current implementation is not the most efficient one. I created [another ticket](https://mattermost.atlassian.net/browse/MM-48469) (hopefully the link is ok now) for tracking this.

**Comment by DonCziken** [comment]
@agarciamontoro below you will find answers to your questions, however  I've also done one more test, basically run the xhr request directly via browser and as a curl. In both cases the response times are about 200-250ms:
![Screenshot from 2022-11-16 23-17-23](https://user-images.githubusercontent.com/36049985/202306713-a3823082-4be7-4ac7-a96f-b88a290842d1.png)

So my guess now is that the slowness maybe caused be other query being run in parallel during the init, however if you look at first SS from issue notification you can see that there are two same requests happening, and the faster one (200ms) starts before the slow one(600ms). I've tried to manually run those two at same time, but did not see any slowdown.

> How many user does that query return for you?

```sql
SELECT COUNT(DISTINCT u.Id) FROM Users AS u WHERE u.DeleteAt = 0;
 count 
-------
    35
(1 row)
```
> Could you also let me know the result of the query?

Sure its roughly 800k, however you have all the numbers (well from yesterday) on the screenshot here: https://user-images.githubusercontent.com/36049985/202002611-484967b1-4855-45d8-b32d-b00c724d1819.png

but just in case here is a result of fresh sql run:

```sql
select count(*) from Posts;
 count  
--------
 798113
(1 row)
```

>Would it be possible for you to time all queries done for that endpoint (you can see them [here](https://github.com/mattermost/mattermost-server/blob/ab5137395cea997f78498caec9bbbe8af224125a/app/analytics.go#L33-L177)) and find out the one that takes the longest?

All of the other queries are extremely fast compared to posts counting:

```sql

#unique_user_count
EXPLAIN ANALYZE SELECT COUNT(DISTINCT u.Id) FROM Users AS u WHERE u.DeleteAt = 0;
                                                QUERY PLAN                                                
----------------------------------------------------------------------------------------------------------
 Aggregate  (cost=7.83..7.83 rows=1 width=8) (actual time=0.077..0.078 rows=1 loops=1)
   ->  Seq Scan on users u  (cost=0.00..7.74 rows=35 width=27) (actual time=0.009..0.028 rows=35 loops=1)
         Filter: (deleteat = 0)
         Rows Removed by Filter: 24
 Planning Time: 2.184 ms
 Execution Time: 0.140 ms
(6 rows)

#team_count
EXPLAIN ANALYZE SELECT COUNT(*) FROM Teams AS t WHERE t.DeleteAt = 0;
                                               QUERY PLAN                                               
--------------------------------------------------------------------------------------------------------
 Aggregate  (cost=1.26..1.27 rows=1 width=8) (actual time=0.029..0.030 rows=1 loops=1)
   ->  Seq Scan on teams t  (cost=0.00..1.26 rows=1 width=0) (actual time=0.019..0.023 rows=16 loops=1)
         Filter: (deleteat = 0)
         Rows Removed by Filter: 5
 Planning Time: 0.153 ms
 Execution Time: 0.072 ms
(6 rows)

#daily_active_users
EXPLAIN ANALYZE SELECT COUNT(*) FROM Status AS s WHERE s.LastActivityAt > 86400000;
                                                QUERY PLAN                                                
----------------------------------------------------------------------------------------------------------
 Aggregate  (cost=2.73..2.74 rows=1 width=8) (actual time=0.054..0.055 rows=1 loops=1)
   ->  Seq Scan on status s  (cost=0.00..2.61 rows=49 width=0) (actual time=0.018..0.045 rows=49 loops=1)
         Filter: (lastactivityat > 86400000)
 Planning Time: 0.570 ms
 Execution Time: 0.095 ms
(5 rows)

#monthly_active_users
EXPLAIN ANALYZE SELECT COUNT(*) FROM Status AS s WHERE s.LastActivityAt > 2678400000;
                                                QUERY PLAN                                                
----------------------------------------------------------------------------------------------------------
 Aggregate  (cost=2.73..2.74 rows=1 width=8) (actual time=0.042..0.043 rows=1 loops=1)
   ->  Seq Scan on status s  (cost=0.00..2.61 rows=49 width=0) (actual time=0.014..0.032 rows=49 loops=1)
         Filter: (lastactivityat > '2678400000'::bigint)
 Planning Time: 0.170 ms
 Execution Time: 0.077 ms
(5 rows)

#inactive_user_count
EXPLAIN ANALYZE SELECT COUNT(Id) FROM Users WHERE DeleteAt > 0;
                                               QUERY PLAN                                               
--------------------------------------------------------------------------------------------------------
 Aggregate  (cost=7.79..7.80 rows=1 width=8) (actual time=0.038..0.039 rows=1 loops=1)
   ->  Seq Scan on users  (cost=0.00..7.74 rows=23 width=27) (actual time=0.009..0.031 rows=24 loops=1)
         Filter: (deleteat > 0)
         Rows Removed by Filter: 35
 Planning Time: 0.280 ms
 Execution Time: 0.097 ms
(6 rows)

#channel_open_count (for default team on our SRV)
EXPLAIN ANALYZE SELECT COUNT(*) AS Value FROM Channels c WHERE c.type = 'O' and c.teamid = '5qe8fpgtetrejmekq6eedeshcy';
                                                                QUERY PLAN                                                                
------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=27.72..27.73 rows=1 width=8) (actual time=0.166..0.167 rows=1 loops=1)
   ->  Bitmap Heap Scan on channels c  (cost=1.79..27.59 rows=53 width=0) (actual time=0.076..0.151 rows=80 loops=1)
         Recheck Cond: (((teamid)::text = '5qe8fpgtetrejmekq6eedeshcy'::text) AND (type = 'O'::channel_type))
         Heap Blocks: exact=23
         ->  Bitmap Index Scan on idx_channels_team_id_type  (cost=0.00..1.78 rows=53 width=0) (actual time=0.029..0.029 rows=83 loops=1)
               Index Cond: (((teamid)::text = '5qe8fpgtetrejmekq6eedeshcy'::text) AND (type = 'O'::channel_type))
 Planning Time: 0.223 ms
 Execution Time: 0.263 ms
(8 rows)

#channel_private_count (for default team on our SRV)
EXPLAIN ANALYZE SELECT COUNT(*) AS Value FROM Channels c WHERE c.type = 'P' and c.teamid = '5qe8fpgtetrejmekq6eedeshcy';
                                                                QUERY PLAN                                                                
------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=22.35..22.36 rows=1 width=8) (actual time=0.162..0.163 rows=1 loops=1)
   ->  Bitmap Heap Scan on channels c  (cost=1.58..22.27 rows=32 width=0) (actual time=0.057..0.144 rows=74 loops=1)
         Recheck Cond: (((teamid)::text = '5qe8fpgtetrejmekq6eedeshcy'::text) AND (type = 'P'::channel_type))
         Heap Blocks: exact=25
         ->  Bitmap Index Scan on idx_channels_team_id_type  (cost=0.00..1.57 rows=32 width=0) (actual time=0.040..0.040 rows=79 loops=1)
               Index Cond: (((teamid)::text = '5qe8fpgtetrejmekq6eedeshcy'::text) AND (type = 'P'::channel_type))
 Planning Time: 0.327 ms
 Execution Time: 0.280 ms
(8 rows)

#channel_open_count
EXPLAIN ANALYZE SELECT COUNT(*) AS Value FROM Channels c WHERE c.type = 'O';
                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=35.42..35.43 rows=1 width=8) (actual time=0.481..0.482 rows=1 loops=1)
   ->  Seq Scan on channels c  (cost=0.00..34.75 rows=270 width=0) (actual time=0.064..0.419 rows=270 loops=1)
         Filter: (type = 'O'::channel_type)
         Rows Removed by Filter: 510
 Planning Time: 0.245 ms
 Execution Time: 0.568 ms
(6 rows)

#channel_private_count
EXPLAIN ANALYZE SELECT COUNT(*) AS Value FROM Channels c WHERE c.type = 'P';
                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=35.15..35.16 rows=1 width=8) (actual time=0.575..0.576 rows=1 loops=1)
   ->  Seq Scan on channels c  (cost=0.00..34.75 rows=160 width=0) (actual time=0.103..0.545 rows=160 loops=1)
         Filter: (type = 'P'::channel_type)
         Rows Removed by Filter: 620
 Planning Time: 0.313 ms
 Execution Time: 0.669 ms
(6 rows)
```

**Comment by hmhealey** [comment]
Thanks for filing the ticket, Alejandro. I took a quick look at the code, and it looks like there's a few places where we're loading the total user count (and possibly some other stats) for things like triggering the cloud trial prompt.

I'll talk to Growth team about possibly changing how that's triggered, but I think we might also want to look at adding a more minimal version of the analytics API since I have to assume that just counting the total number of users wouldn't take that long if we even do need it on initial load.

**Comment by agarciamontoro** [comment]
Thank you for your detailed response and sorry for the delay, @DonCziken. It's interesting that this seems to happen only when all queries are done at the same time. Can you reproduce that situation manually, maybe with a script?

In any case, I have some good news to share: the Growth team already opened a PR to make these calls only for admins :) https://github.com/mattermost/mattermost-webapp/pull/11735

**Comment by DonCziken** [comment]
@agarciamontoro  great to hear that FE will be updated! :)

I will try to replicate the lag with some scripting, but will only have time to look at this next week, so stay tuned!
